src/ProxyManager/Generator/MagicMethodGenerator.php

```diff
@@ -31,24 +33,55 @@
 class MagicMethodGenerator extends MethodGenerator
 {
     /**
-     * @param ReflectionClass $originalClass
-     * @param string          $name
-     * @param array           $parameters
+     * @param ReflectionClass      $originalClass
+     * @param string               $name
+     * @param ParameterGenerator[] $parameters
+     *
+     * @throws \Zend\Code\Generator\Exception\InvalidArgumentException
      */
     public function __construct(ReflectionClass $originalClass, string $name, array $parameters = [])
     {
-        parent::__construct(
-            $name,
-            $parameters,
-            static::FLAG_PUBLIC,
-            null,
-            $originalClass->hasMethod($name) ? '{@inheritDoc}' : null
-        );
+        parent::__construct($name, [], static::FLAG_PUBLIC);
 
         $this->setReturnsReference(strtolower($name) === '__get');
 
         if ($originalClass->hasMethod($name)) {
-            $this->setReturnsReference($originalClass->getMethod($name)->returnsReference());
+            $originalMethod = $originalClass->getMethod($name);
+
+            $this->setDocBlock('{@inheritDoc}');
+            $this->setReturnsReference($originalMethod->returnsReference());
+            $this->setReturnType(((string) $originalMethod->getReturnType()) ?: null);
+            $this->addOrReplaceParameterNames($originalMethod, ...$parameters);
+        } else {
+            $this->setParameters($parameters);
         }
     }
+
+    private function addOrReplaceParameterNames(
```

Comment:
This method is way too complex: need to split into multiple private methods explaining exactly what is going on at each step.